File: mapstruct/mapstruct-java-extension/src/main/java/io/koraframework/mapstruct/java/extension/MapstructKoraExtensionFactory.java
package io.koraframework.mapstruct.java.extension;

import io.koraframework.kora.app.annotation.processor.extension.ExtensionFactory;
import io.koraframework.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static io.koraframework.mapstruct.java.extension.MapstructKoraExtension.MAPPER_ANNOTATION;

public final class MapstructKoraExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var mapper = processingEnvironment.getElementUtils().getTypeElement(MAPPER_ANNOTATION.canonicalName());
        if (mapper == null) {
            return Optional.empty();
        }
        return Optional.of(new MapstructKoraExtension(processingEnvironment));
    }
}

File: mapstruct/mapstruct-ksp-extension/src/main/kotlin/io/koraframework/mapstruct/ksp/extension/MapstructKoraExtensionFactory.kt
package io.koraframework.mapstruct.ksp.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import io.koraframework.kora.app.ksp.extension.ExtensionFactory
import io.koraframework.kora.app.ksp.extension.KoraExtension

class MapstructKoraExtensionFactory : ExtensionFactory {

    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        return resolver.getClassDeclarationByName(MapstructKoraExtension.mapperAnnotation.canonicalName)
            ?.let { MapstructKoraExtension }
    }
}
